src/main.rs

```diff
@@ -102,82 +117,92 @@ impl Author {
 }
 
 fn main() -> Result<(), Box<Error>> {
-    let opt = Cli::from_args();
-    let curses: Vec<&str> = CURSES.lines().collect();
-    let path = if opt.directory.is_none() {
-        env::current_dir()?
-    } else {
-        opt.directory.unwrap()
+    let repo_path = {
+        let opt = Cli::from_args();
+        if let Some(directory) = opt.directory {
+            directory
+        } else {
+            env::current_dir()?
+        }
     };
-    let repo = Repository::open(&path)?;
-    let mut revwalk = repo.revwalk()?;
-    let mut commits: Vec<Commit> = Vec::new();
-    revwalk.push_head()?;
-    for commit in revwalk {
-        let commit = repo.find_commit(commit?)?;
-        commits.push(commit);
-    }
-    let mut repo = Repo::new(path.file_name().unwrap().to_str().unwrap());
-    let mut authors: Vec<Author> = find_authors(&commits);
+
+    let repo = Repository::open(&repo_path)?;
+    let commits = {
+        let mut revwalk = repo.revwalk()?;
+        let mut commits = Vec::new();
+        revwalk.push_head()?;
+        for commit_id in revwalk {
+            let commit = repo.find_commit(commit_id?)?;
+            commits.push(commit);
+        }
+        commits
+    };
+
+    let (s, r) = crossbeam_channel::unbounded();
+
+    let mut repo = Repo::new(repo_path.file_name().unwrap().to_str().unwrap());
     for commit in &commits {
-        let text = commit.message().unwrap().to_lowercase().to_string();
-        let author = commit.author().name().unwrap().to_string();
-        let index = authors
-            .iter()
-            .position(|i| i.name == author)
-            .expect("Could not find author");
-        let mut author = &mut authors[index];
-        author.total_commits += 1;
-        repo.total_commits += 1;
-        for word in text.split_whitespace() {
-            let word = clean_word(word);
-            if naughty_word(word.as_str(), &curses) {
-                author.total_curses += 1;
-                repo.total_curses += 1;
-                author.update_occurrence(word.as_str());
+        if let (Some(author_name), Some(commit_message)) = (
+            commit.author().name().map(|n| n.to_owned()),
+            commit.message().map(|msg| msg.to_lowercase()),
+        ) {
+            let job_sender = s.clone();
+            rayon::spawn(move || {
+                let mut naughty_words = HashMap::new();
+                for word in split_into_clean_words(&commit_message).filter(|w| naughty_word(w)) {
+                    naughty_words
+                        .get_mut(word)
+                        .map(|count| *count += 1)
+                        .unwrap_or_else(|| {
+                            naughty_words.insert(word.to_owned(), 1);
```

Comment:
The change from L111 could also be applied here

Reply:
See previous comment ^, avoiding string copies unless necessary.

I would like to find a crate that extends the `HashMap<String, T>` API to clean up the duplication though..